aten/src/ATen/native/DistributionTemplates.h

```diff
@@ -187,58 +187,94 @@ Tensor& normal_impl_(Tensor& self, double mean, double std, c10::optional<Genera
-  output.add_(mean);
+  CHECK_NORMAL_STD(std);
+  if (output.is_meta()) {
+    auto std_tensor = at::empty_like(output, c10::kMeta, MemoryFormat::Contiguous);
+    auto shape = at::infer_size(mean.sizes(), std_tensor.sizes());
+    at::native::resize_output(output, shape);
+  } else {
+    normal_impl_<normal_kernel, RNG>(output, 0, std, gen);
+    output.add_(mean);
+  }
   return output;
 }
 
 template<template<typename> class normal_kernel, typename RNG>
 Tensor& normal_out_impl(Tensor& output, double mean, const Tensor& std, c10::optional<Generator> gen) {
   CHECK_NORMAL_TENSOR_STD(std);
-  normal_impl_<normal_kernel, RNG>(output, 0, 1, gen);
-  auto mean_tensor = at::full({}, mean, output.options());
-  // CUDA NB: addcmul_out copies the tensor to be added into the output.
-  // Please look at aten/src/THC/generic/THCTensorMathPointwise.cu
-  // The previous function here was addcmul_out(output, mean_tensor, output, std, 1);
-  // The third argument is not a constant reference and hence the samples in output are overwritten.
-  // Consequently, the computation performed is mean_tensor + mean_tensor * std instead of mean_tensor + output * std
-  output.mul_(std).add_(mean_tensor);
+  if (output.is_meta()) {
+    auto mean_tensor = at::empty_like(output, c10::kMeta, MemoryFormat::Contiguous);
+    auto shape = at::infer_size(mean_tensor.sizes(), std.sizes());
+    at::native::resize_output(output, shape);
+  } else {
+    normal_impl_<normal_kernel, RNG>(output, 0, 1, gen);
+    auto mean_tensor = at::full({}, mean, output.options());
+    // CUDA NB: addcmul_out copies the tensor to be added into the output.
+    // Please look at aten/src/THC/generic/THCTensorMathPointwise.cu
+    // The previous function here was addcmul_out(output, mean_tensor, output, std, 1);
+    // The third argument is not a constant reference and hence the samples in output are overwritten.
+    // Consequently, the computation performed is mean_tensor + mean_tensor * std instead of mean_tensor + output * std
+    output.mul_(std).add_(mean_tensor);
+  }
   return output;
 }
 
 template<template<typename> class normal_kernel, typename RNG>
 Tensor& normal_out_impl(Tensor& output, const Tensor& mean, const Tensor& std, c10::optional<Generator> gen) {
   CHECK_NORMAL_TENSOR_STD(std);
-  auto shape = at::infer_size(mean.sizes(), std.sizes());
-  at::native::resize_output(output, shape);
-  normal_impl_<normal_kernel, RNG>(output, 0, 1, gen);
-  // CUDA NB: addcmul_out copies the tensor to be added into the output.
-  // Please look at aten/src/THC/generic/THCTensorMathPointwise.cu
-  // The previous function here was addcmul_out(output, mean, output, std, 1);
-  // The third argument is not a constant reference and hence the samples in output are overwritten.
-  // Consequently, the computation performed is mean + mean * std instead of mean + output * std
-  output.mul_(std).add_(mean);
+  if (output.is_meta()) {
```

Comment:
I don't think I'm following the changes here. Why do you need to special case `is_meta()`?

Right now if output `is_meta()` is true, we'll skip the `normal_impl<...>()` call and the `output.mul_(std).add_(mean)`. But those should be fine to run with meta tensors, since they'll just no-op.

If `output.is_meta()` is false, then we'll end up skipping the resize in the if statement, which seems like a behavior change from before.

Reply:
oh, is this fixed in the next PR (https://github.com/pytorch/pytorch/pull/72004/files) ?

Reply:
If you're ok with merging these all into a single PR, I think that might make the stack easier to review :). ghstack will by default create a separate PR for every commit in your branch. So if you merge them all together (or amend changes to your existing commits instead of creating new ones), then you can keep logical changes in the same PR.

Reply:
@bdhirsh oh, sorry about that! my idea for doing it this way was to show the changes i made in a step by step fashion. let me squash the relevant commits, then! will ping again once it's done